File: ts-node/src/engine.ts
export enum Side { Buy, Sell }
export enum OrdType { Limit, Market }

export interface Order {
  id: number; side: Side; typ: OrdType; price: number; qty: number; ts: number;
}

type Level = { price:number, q: Order[] };

export class OrderBook {
  asks: Level[] = []; // ascending
  bids: Level[] = []; // descending
  nextTs = 0;
  trades = 0; filledOrders = 0; filledQty = 0;

  private pushLevel(levels: Level[], price:number, o:Order, asc:boolean) {
    const idx = levels.findIndex(l => l.price===price);
    if (idx>=0) { levels[idx].q.push(o); return; }
    levels.push({price, q:[o]});
    levels.sort((a,b)=> asc ? a.price-b.price : b.price-a.price);
  }

  private matchAgainst(levels: Level[], isAskBook:boolean, incoming: Order): number {
    let rem = incoming.qty;
    while (rem>0 && levels.length>0) {
      const best = levels[0];
      const priceOk = incoming.typ===OrdType.Market ||
        (incoming.side===Side.Buy && best.price<=incoming.price) ||
        (incoming.side===Side.Sell && best.price>=incoming.price);
      if (!priceOk) break;
      while (rem>0 && best.q.length>0) {
        const head = best.q[0];
        const trade = Math.min(rem, head.qty);
        rem -= trade; head.qty -= trade;
        this.trades++; this.filledQty += trade;
        if (head.qty===0) { best.q.shift(); this.filledOrders++; } else { break; }
      }
      if (best.q.length===0) { levels.shift(); }
    }
    return rem;
  }

  submit(o: Order) {
    this.nextTs++; o.ts = this.nextTs;
    if (o.side===Side.Buy) {
      const rem = this.matchAgainst(this.asks, true, o);
      if (rem>0 && o.typ===OrdType.Limit) { o.qty=rem; this.pushLevel(this.bids, o.price, o, false); }
    } else {
      const rem = this.matchAgainst(this.bids, false, o);
      if (rem>0 && o.typ===OrdType.Limit) { o.qty=rem; this.pushLevel(this.asks, o.price, o, true); }
    }
  }
}

File: go/engine/engine.go
package engine

import (
	"slices"
)

type Side int
const (
	Buy Side = iota
	Sell
)

type OrdType int
const (
	Limit OrdType = iota
	Market
)

type Order struct {
	Id    uint64
	Side  Side
	Type  OrdType
	Price int64
	Qty   int64
	Ts    uint64
}

type level struct {
	price int64
	q     []Order // FIFO
}

type OrderBook struct {
	asks []level // ascending price
	bids []level // descending price
	nextTs uint64

	Trades uint64
	FilledOrders uint64
	FilledQty int64
}

func NewOrderBook() *OrderBook {
	return &OrderBook{}
}

func (ob *OrderBook) pushLevel(levels *[]level, price int64, o Order, asc bool) {
	// find price level
	idx := slices.IndexFunc(*levels, func(l level) bool { return l.price == price })
	if idx >= 0 {
		(*levels)[idx].q = append((*levels)[idx].q, o)
		return
	}
	// insert keeping order
	l := level{price: price, q: []Order{o}}
	*levels = append(*levels, l)
	if asc {
		slices.SortFunc(*levels, func(a, b level) int {
			if a.price < b.price { return -1 }
			if a.price > b.price { return 1 }
			return 0
		})
	} else {
		slices.SortFunc(*levels, func(a, b level) int {
			if a.price > b.price { return -1 }
			if a.price < b.price { return 1 }
			return 0
		})
	}
}

func (ob *OrderBook) matchAgainst(levels *[]level, isAskBook bool, incoming Order) int64 {
	rem := incoming.Qty
	for rem > 0 && len(*levels) > 0 {
		// best price at index 0 due to sorting
		best := &(*levels)[0]
		priceOk := incoming.Type == Market ||
			(incoming.Side == Buy && best.price <= incoming.Price) ||
			(incoming.Side == Sell && best.price >= incoming.Price)
		if !priceOk { break }

		// consume FIFO
		for rem > 0 && len(best.q) > 0 {
			head := best.q[0]
			trade := head.Qty
			if trade > rem { trade = rem }
			rem -= trade
			head.Qty -= trade
			ob.Trades++
			ob.FilledQty += trade
			if head.Qty > 0 {
				best.q[0] = head
				break
			} else {
				ob.FilledOrders++
				best.q = best.q[1:]
			}
		}
		if len(best.q) == 0 {
			// remove level
			*levels = (*levels)[1:]
		}
	}
	return rem
}

func (ob *OrderBook) Submit(o Order) {
	ob.nextTs++
	o.Ts = ob.nextTs
	if o.Side == Buy {
		rem := ob.matchAgainst(&ob.asks, true, o)
		if rem > 0 && o.Type == Limit {
			o.Qty = rem
			ob.pushLevel(&ob.bids, o.Price, o, false)
		}
	} else {
		rem := ob.matchAgainst(&ob.bids, false, o)
		if rem > 0 && o.Type == Limit {
			o.Qty = rem
			ob.pushLevel(&ob.asks, o.Price, o, true)
		}
	}
}
